gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleHelixTask.java

```diff
@@ -0,0 +1,149 @@
+      TaskAttemptBuilder taskAttemptBuilder, StateStores stateStores) {
+    this.jobId = jobId;
+    this.workUnitFilePath = workUnitFilePath;
+    this.jobStateFilePath = jobStateFilePath;
+    this.fs = fs;
+    this.taskAttemptBuilder = taskAttemptBuilder;
+    this.stateStores = stateStores;
+  }
+
+  public int run()
+      throws IOException, InterruptedException {
+    SharedResourcesBroker<GobblinScopeTypes> globalBroker = null;
+    try {
+      return runInternal(globalBroker);
+    }finally {
+      closeGlobalBroker(globalBroker);
+    }
+  }
+
+  private int runInternal(SharedResourcesBroker<GobblinScopeTypes> globalBroker)
+      throws IOException, InterruptedException {
+    List<WorkUnit> workUnits = getWorkUnits();
+    int workUnitSize = workUnits.size();
+
+    JobState jobState = getJobState();
+
+    globalBroker = SharedResourcesBrokerFactory
+        .createDefaultTopLevelBroker(ConfigFactory.parseProperties(jobState.getProperties()), GobblinScopeTypes.GLOBAL.defaultScopeInstance());
+    SharedResourcesBroker<GobblinScopeTypes> jobBroker =
+        globalBroker.newSubscopedBuilder(new JobScopeInstance(jobState.getJobName(), jobState.getJobId())).build();
+
+    this.taskAttempt = this.taskAttemptBuilder.build(workUnits.iterator(), this.jobId, jobState, jobBroker);
+    this.taskAttempt.runAndOptionallyCommitTaskAttempt(GobblinMultiTaskAttempt.CommitPolicy.IMMEDIATE);
+    return workUnitSize;
+
+  }
+  private void closeGlobalBroker(SharedResourcesBroker<GobblinScopeTypes> globalBroker) {
+    if (globalBroker != null) {
+      try {
+        globalBroker.close();
+      } catch (IOException ioe) {
+        _logger.error("Could not close shared resources broker.", ioe);
+      }
+    }
+  }
+
+  private JobState getJobState()
+      throws java.io.IOException {
+    JobState jobState = new JobState();
+    SerializationUtils.deserializeState(this.fs, jobStateFilePath, jobState);
+    return jobState;
+  }
+
+  private List<WorkUnit> getWorkUnits()
+      throws IOException {
+    String fileName = workUnitFilePath.getName();
+    String storeName = workUnitFilePath.getParent().getName();
+    WorkUnit workUnit;
+
+    if (workUnitFilePath.getName().endsWith(AbstractJobLauncher.MULTI_WORK_UNIT_FILE_EXTENSION)) {
+      workUnit = stateStores.mwuStateStore.getAll(storeName, fileName).get(0);
+    } else {
+      workUnit = stateStores.wuStateStore.getAll(storeName, fileName).get(0);
+    }
+
+    // The list of individual WorkUnits (flattened) to run
+    List<WorkUnit> workUnits = Lists.newArrayList();
+
+    if (workUnit instanceof MultiWorkUnit) {
+      // Flatten the MultiWorkUnit so the job configuration properties can be added to each individual WorkUnits
+      List<WorkUnit> flattenedWorkUnits = JobLauncherUtils.flattenWorkUnits(((MultiWorkUnit) workUnit).getWorkUnits());
+      workUnits.addAll(flattenedWorkUnits);
```

Comment:
Is it possible to simplify this? Maybe return flattenedWorkunits directly in if condition? In else condition just return ImmutableList.of(workunit)?

Reply:
Maybe,
I just moved the existing code. I would prefer not to make this change for this PR.